### PR TITLE
[warm reboot] Warm Reboot Support for EVPN VXLAN

### PR DESCRIPTION
**Pull request by srj102**
1. Added new enums to WarmStartState
2. Added a new function to get the warm start state.

Refer to HLD: https://github.com/Azure/SONiC/pull/437

### CODE REVIEW

**Comment by prsunny** [comment]
Request review from @dzhangalibaba , @zhenggen-xu

File: common/warm_restart.cpp [hunk 1]
```diff
@@ -181,6 +181,41 @@ bool WarmStart::isSystemWarmRebootEnabled(void)
     return warmStart.m_systemWarmRebootEnabled;
 }
 
+void WarmStart::getWarmStartState(const std::string &app_name, WarmStartState &state)
+{
+    //WarmStartStateNameMap::iterator it;
```
Comment (qiluo-msft) [inline common/warm_restart.cpp:186]: >//WarmStartStateNameMap::iterator it; [](start = 4, length = 37)

Remove commented code

File: common/warm_restart.cpp [hunk 2]
```diff
@@ -181,6 +181,41 @@ bool WarmStart::isSystemWarmRebootEnabled(void)
     return warmStart.m_systemWarmRebootEnabled;
 }
 
+void WarmStart::getWarmStartState(const std::string &app_name, WarmStartState &state)
```
Comment (qiluo-msft) [inline common/warm_restart.cpp:184]: >WarmStartState [](start = 63, length = 14)

Why not just return this state?

File: common/warm_restart.cpp [hunk 3]
```diff
@@ -181,6 +181,41 @@ bool WarmStart::isSystemWarmRebootEnabled(void)
...
+      return;
+
+    warmStart.m_stateWarmRestartTable->hget(app_name, "state", statestr);
+
+    /* If warm-start is enabled, state cannot be assumed as Reconciled
+     * It should be set to unknown
+     */
+    state = WSUNKNOWN;
+
+    for(std::map<WarmStart::WarmStartState, std::string>::const_iterator it = warmStartStateNameMap.begin(); it != warmStartStateNameMap.end(); it++)
```
Comment (qiluo-msft) [inline common/warm_restart.cpp:203]: >std::map<WarmStart::WarmStartState, std::string>::const_iterator [](start = 8, length = 64)

auto?

File: common/warm_restart.cpp [hunk 4]
```diff
@@ -181,6 +181,41 @@ bool WarmStart::isSystemWarmRebootEnabled(void)
...
+    {
+      if(it->second == statestr)
+      {
+         state = it->first;
+         break;
+      }
```
Comment (qiluo-msft) [inline common/warm_restart.cpp:209]: >   [](start = 4, length = 2)

Use 4x spaces as indentation

File: common/warm_restart.cpp [hunk 5]
```diff
@@ -181,6 +181,41 @@ bool WarmStart::isSystemWarmRebootEnabled(void)
...
+
+    if(!isWarmStart())
+      return;
+
+    warmStart.m_stateWarmRestartTable->hget(app_name, "state", statestr);
+
+    /* If warm-start is enabled, state cannot be assumed as Reconciled
+     * It should be set to unknown
+     */
+    state = WSUNKNOWN;
```
Comment (qiluo-msft) [inline common/warm_restart.cpp:201]: >state = WSUNKNOWN; [](start = 4, length = 18)

Is it expected state in runtime? If not, just throw an exception.

File: common/warm_restart.cpp (hunk 2 above)
Comment (nkelapur) [reply common/warm_restart.cpp:184]: Don't see any issue with pass by reference. Will consider that while updating other files where this api is used.

File: common/warm_restart.cpp (hunk 1 above)
Comment (nkelapur) [reply common/warm_restart.cpp:186]: Will remove

File: common/warm_restart.cpp (hunk 5 above)
Comment (nkelapur) [reply common/warm_restart.cpp:201]: Yes, this is expected. When Module A check for the status of Module B, module B may not have reached/initialised to any state yet. Hence module A needs to know that Module B is not ready yet an act accordingly.

File: common/warm_restart.cpp (hunk 3 above)
Comment (nkelapur) [reply common/warm_restart.cpp:203]: will update

File: common/warm_restart.cpp (hunk 4 above)
Comment (nkelapur) [reply common/warm_restart.cpp:209]: will change

File: common/warm_restart.cpp [hunk 6]
```diff
@@ -181,6 +184,39 @@ bool WarmStart::isSystemWarmRebootEnabled(void)
...
 
+void WarmStart::getWarmStartState(const std::string &app_name, WarmStartState &state)
+{
+    std::string statestr;
+
+    auto& warmStart = getInstance();
+
+    state = RECONCILED;
+
+    if(!isWarmStart())
```
Comment (prsunny) [inline common/warm_restart.cpp:195]: Please have a space after `if`, also with open/close braces. Please do it for the rest of the `if `and `for `below

Comment (nkelapur) [reply common/warm_restart.cpp]: Done.

**Review by prsunny — APPROVED** [review]

**Review by qiluo-msft — APPROVED** [review]

**Pull request merged**